nbclient/execute.py

```diff
@@ -502,14 +532,41 @@ def _passed_deadline(self, deadline):
         return False
 
     def run_cell(self, cell, cell_index=0, store_history=False):
+        loop = get_loop()
+        return loop.run_until_complete(self.async_run_cell(cell, cell_index, store_history))
+
+    async def poll_output_msg(self, parent_msg_id, cell, cell_index, timeout=None):
```

Comment:
Will need to do some more manual testing to ensure we correctly timeout, pass execution context, and handle errors cleanly. Some of those patterns are tested in the synchronous path with some assumptions that might not hold for the async version.

Reply:
Just to explain a bit more how I implemented the asynchronous functionality, there are now two tasks that are run in parallel:
- `_poll_for_reply`: asynchronously awaits a reply on the shell channel (possibly with timeout).
- `poll_output_msg`: asynchronously awaits a message on the IOPub channel for cell execution. We first await as long as no reply is received on the shell channel, and when a reply is received on the shell channel we cancel this task and launch it again with a timeout.

Reply:
Thanks, that helps!